set: add pmemset_remove_range function

kswiecicki opened the pull request:
In accordance with the discussed changes it replaces https://github.com/pmem/pmdk/pull/5175.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/pmdk/5185)
<!-- Reviewable:end -->

pbalcer reviewed (changes requested):
> Reviewed 5 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmdk/5185)* status: 5 of 9 files reviewed, 8 unresolved discussions (waiting on @kswiecicki)

___
*[src/libpmemset/part.c, line 267 at r1](https://reviewable.io/reviews/pmem/pmdk/5185#-MW3sK0l8uQnq-rFKwSz:-MW3sK0l8uQnq-rFKwT-:b-cod5zn) ([raw file](https://github.com/pmem/pmdk/blob/e911f2fb116803d0f9d1cdf2dca3cda166da28aa/src/libpmemset/part.c#L267)):*
> ```C
>  */
> void
> pmemset_part_map_delete(struct pmemset_part_map **part_map)
> ```

A "delete" function should literally deinitialize the stuff inside of a struct and free its memory.

This isn't that.
___
*[src/libpmemset/part.c, line 288 at r1](https://reviewable.io/reviews/pmem/pmdk/5185#-MW3tzJcDAH06l55krxN:-MW3tzJcDAH06l55krxO:bx73s1z) ([raw file](https://github.com/pmem/pmdk/blob/e911f2fb116803d0f9d1cdf2dca3cda166da28aa/src/libpmemset/part.c#L288)):*
> ```C
> 		/* pmem2 mapping covers whole vm reservation */
> 		pmem2_vm_reservation_delete(&pmem2_reserv);
> 		Free(pmap);
> ```

why is this memory only conditionally freed?

This kind of logic should be outside of a "delete", or shouldn't be conditional.
___
*[src/libpmemset/part.c, line 301 at r1](https://reviewable.io/reviews/pmem/pmdk/5185#-MW3tdhY1LWaCdQZRR2g:-MW3tdhY1LWaCdQZRR2h:b-vt0p0) ([raw file](https://github.com/pmem/pmdk/blob/e911f2fb116803d0f9d1cdf2dca3cda166da28aa/src/libpmemset/part.c#L301)):*
> ```C
> /*
>  * pmemset_part_map_delete_with_contents -- removes the part map structure and
>  *                                          unmaps underlying pmem2 mappings
> ```

I don't get it - doesn't the function above do the exact same thing?
(I know it doesn't by looking at code - but you can parametrize this function to just remove the earliest pmem2_map).

You should always try to minimize special cases - if you have multiple similar operations, think of a way to represent all of them using single common logic.
___
*[src/libpmemset/part.c, line 313 at r1](https://reviewable.io/reviews/pmem/pmdk/5185#-MW3tn-XDKsTPJmJO72U:-MW3tn-XDKsTPJmJO72V:bqf1j4s) ([raw file](https://github.com/pmem/pmdk/blob/e911f2fb116803d0f9d1cdf2dca3cda166da28aa/src/libpmemset/part.c#L313)):*
> ```C
> 	struct pmem2_map *map;
> 	pmem2_vm_reservation_map_find(rsv, 0, rsv_size, &map);
> 	while (map) {
> ```

you can easily use a single map find...
___
*[src/libpmemset/part.c, line 325 at r1](https://reviewable.io/reviews/pmem/pmdk/5185#-MW3s67A5hTLBrTuTHK5:-MW3s67BAILEDNtzCo83:b-yqxl38) ([raw file](https://github.com/pmem/pmdk/blob/e911f2fb116803d0f9d1cdf2dca3cda166da28aa/src/libpmemset/part.c#L325)):*
> ```C
> 		return ret;
> 
> 	Free(pmap);
> ```

there should typically be only one "free" that matches one "malloc". I counted three different places where we are deallocating a part map.
___
*[src/libpmemset/part.c, line 347 at r1](https://reviewable.io/reviews/pmem/pmdk/5185#-MW3t3kXEAIxQpdSYPqH:-MW3t3kXEAIxQpdSYPqI:b-p4f4w2) ([raw file](https://github.com/pmem/pmdk/blob/e911f2fb116803d0f9d1cdf2dca3cda166da28aa/src/libpmemset/part.c#L347)):*
> ```C
> 	struct pmem2_map *map;
> 	pmem2_vm_reservation_map_find(pmem2_rsv, offset, size, &map);
> 	while (map) {
> ```

can you reorganize this loop so that only one map_find is used?
This looks overly complicated to me.
___
*[src/libpmemset/pmemset.c, line 519 at r1](https://reviewable.io/reviews/pmem/pmdk/5185#-MW3vmj9AuLLMp0rVH-J:-MW3vmj9AuLLMp0rVH-K:b4ahkrx) ([raw file](https://github.com/pmem/pmdk/blob/e911f2fb116803d0f9d1cdf2dca3cda166da28aa/src/libpmemset/pmemset.c#L519)):*
> ```C
>  */
> static int
> pmemset_remove_part_map_range(struct pmemset *set,
> ```

pmemset_delete is a special case of this function.

Think if there's a way to generalize this whole map delete/remove problem.
___
*[src/libpmemset/pmemset.c, line 554 at r1](https://reviewable.io/reviews/pmem/pmdk/5185#-MW3wwVJ8mNUndfp9vKm:-MW3wwVJ8mNUndfp9vKn:b9aovum) ([raw file](https://github.com/pmem/pmdk/blob/e911f2fb116803d0f9d1cdf2dca3cda166da28aa/src/libpmemset/pmemset.c#L554)):*
> ```C
>  */
> int
> pmemset_remove_range(struct pmemset *set, void *addr, size_t len)
> ```

what happens if this removes a whole part_map? does it leak memory?


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[src/libpmemset/part.c, line 267 at r1](https://reviewable.io/reviews/pmem/pmdk/5185#-MW3sK0l8uQnq-rFKwSz:-MWdF9fLA8R3tMkXaDNS:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/e911f2fb116803d0f9d1cdf2dca3cda166da28aa/src/libpmemset/part.c#L267)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

A "delete" function should literally deinitialize the stuff inside of a struct and free its memory.

This isn't that.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[src/libpmemset/part.c, line 288 at r1](https://reviewable.io/reviews/pmem/pmdk/5185#-MW3tzJcDAH06l55krxN:-MWdFBgt5I_wY1Fx3Ahx:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/e911f2fb116803d0f9d1cdf2dca3cda166da28aa/src/libpmemset/part.c#L288)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

why is this memory only conditionally freed?

This kind of logic should be outside of a "delete", or shouldn't be conditional.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[src/libpmemset/part.c, line 301 at r1](https://reviewable.io/reviews/pmem/pmdk/5185#-MW3tdhY1LWaCdQZRR2g:-MWdFDbVBZdkjomc2VC0:bzbmk4x) ([raw file](https://github.com/pmem/pmdk/blob/e911f2fb116803d0f9d1cdf2dca3cda166da28aa/src/libpmemset/part.c#L301)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

I don't get it - doesn't the function above do the exact same thing?
(I know it doesn't by looking at code - but you can parametrize this function to just remove the earliest pmem2_map).

You should always try to minimize special cases - if you have multiple similar operations, think of a way to represent all of them using single common logic.
</blockquote></details>

Done. You were right it was useless.


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[src/libpmemset/part.c, line 313 at r1](https://reviewable.io/reviews/pmem/pmdk/5185#-MW3tn-XDKsTPJmJO72U:-MWdFHu65iqTbujzJToX:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/e911f2fb116803d0f9d1cdf2dca3cda166da28aa/src/libpmemset/part.c#L313)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

you can easily use a single map find...
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[src/libpmemset/part.c, line 325 at r1](https://reviewable.io/reviews/pmem/pmdk/5185#-MW3s67A5hTLBrTuTHK5:-MWdFRnd9eGPYu3IGctF:b5qep7n) ([raw file](https://github.com/pmem/pmdk/blob/e911f2fb116803d0f9d1cdf2dca3cda166da28aa/src/libpmemset/part.c#L325)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

there should typically be only one "free" that matches one "malloc". I counted three different places where we are deallocating a part map.
</blockquote></details>

I've tidied it up. Done.


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[src/libpmemset/part.c, line 347 at r1](https://reviewable.io/reviews/pmem/pmdk/5185#-MW3t3kXEAIxQpdSYPqH:-MWdFWE60GnwaKd4C4_X:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/e911f2fb116803d0f9d1cdf2dca3cda166da28aa/src/libpmemset/part.c#L347)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

can you reorganize this loop so that only one map_find is used?
This looks overly complicated to me.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

pbalcer reviewed (changes requested):
looks much nicer now.

> Reviewed 1 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmdk/5185#-:-MWe0Fh3EXHkDEQsL-la:bbq4zt1)* status: 4 of 12 files reviewed, 4 unresolved discussions (waiting on @kswiecicki and @pbalcer)

___
*[src/libpmemset/pmemset.c, line 385 at r2](https://reviewable.io/reviews/pmem/pmdk/5185#-MWe-hrIEEpzYR3tSwop:-MWe-hrIEEpzYR3tSwoq:b-9rmsz6) ([raw file](https://github.com/pmem/pmdk/blob/118a1f445f0d70b2a3b3ec0f41fe8ebd5332fa6d/src/libpmemset/pmemset.c#L385)):*
> ```C
> 	struct pmemset_part_map *pmap = set->previous_pmap;
> 	/* if there are no parts in pmemset, then new part can't be coalesced */
> 	enum pmemset_coalescing coalescing = (pmap) ? set->part_coalescing :
> ```

you already have to handle the case when coalescing is not possible, so it might be possible to remove this condition and handle this condition in one place.
___
*[src/libpmemset/pmemset.c, line 392 at r2](https://reviewable.io/reviews/pmem/pmdk/5185#-MWe-ROz6aeNBR8cqImx:-MWe-ROz6aeNBR8cqImy:b-jfujbg) ([raw file](https://github.com/pmem/pmdk/blob/118a1f445f0d70b2a3b3ec0f41fe8ebd5332fa6d/src/libpmemset/pmemset.c#L392)):*
> ```C
> 			ret = pmemset_part_map_extend_end(pmap, part_size);
> 
> 			if (!ret || coalescing == PMEMSET_COALESCING_FULL)
> ```

if coalescing is opportunistic, and this failed, doesn't this mean that "coalesced" should be false?
(add a comment if that's not the case)


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[src/libpmemset/pmemset.c, line 385 at r2](https://reviewable.io/reviews/pmem/pmdk/5185#-MWe-hrIEEpzYR3tSwop:-MX1bkE1B-O9vphSXN4c:b-d5cque) ([raw file](https://github.com/pmem/pmdk/blob/118a1f445f0d70b2a3b3ec0f41fe8ebd5332fa6d/src/libpmemset/pmemset.c#L385)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

you already have to handle the case when coalescing is not possible, so it might be possible to remove this condition and handle this condition in one place.
</blockquote></details>

One of the reasons I've moved the "coalescing not possible" case out of the switch is to avoid those awkward indents when function parameters are split across lines.
Anyway I've moved it back to the switch, it's more logical that way.


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[src/libpmemset/pmemset.c, line 392 at r2](https://reviewable.io/reviews/pmem/pmdk/5185#-MWe-ROz6aeNBR8cqImx:-MX1cAwd48qh9FRHSdaA:bgjsn6l) ([raw file](https://github.com/pmem/pmdk/blob/118a1f445f0d70b2a3b3ec0f41fe8ebd5332fa6d/src/libpmemset/pmemset.c#L392)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

if coalescing is opportunistic, and this failed, doesn't this mean that "coalesced" should be false?
(add a comment if that's not the case)
</blockquote></details>

If coalescing is opportunistic and the *extend* fails, then it will not break out of the switch(...) but continue in the *PMEMSET_COALESCING_NONE* case, the "coalesced" value changes there. I've added the comment. 


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[src/libpmemset/pmemset.c, line 519 at r1](https://reviewable.io/reviews/pmem/pmdk/5185#-MW3vmj9AuLLMp0rVH-J:-MX1d-fkErdhERutr_1e:b-brpkzw) ([raw file](https://github.com/pmem/pmdk/blob/e911f2fb116803d0f9d1cdf2dca3cda166da28aa/src/libpmemset/pmemset.c#L519)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

pmemset_delete is a special case of this function.

Think if there's a way to generalize this whole map delete/remove problem.
</blockquote></details>

*pmemset_delete* and *pmemset_remove_range* share the same functions now.


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[src/libpmemset/pmemset.c, line 554 at r1](https://reviewable.io/reviews/pmem/pmdk/5185#-MW3wwVJ8mNUndfp9vKm:-MX1dV2Q7Re-FjeSJWvt:b-ru7ccx) ([raw file](https://github.com/pmem/pmdk/blob/e911f2fb116803d0f9d1cdf2dca3cda166da28aa/src/libpmemset/pmemset.c#L554)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

what happens if this removes a whole part_map? does it leak memory?
</blockquote></details>

I omitted resizing/deleting *part_map* at the time and while it didn't leak memory, *ravl* in *pmemset* could contain empty, unmapped part_map records.
*part_map*s should be correctly resized/deleted now.


<!-- Sent from Reviewable.io -->

pbalcer reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmdk/5185#-:-MX6xL6gCxdsYupn9-W4:bnfp4nl)* status: 3 of 12 files reviewed, 1 unresolved discussion (waiting on @kswiecicki and @pbalcer)

___
*[src/libpmemset/pmemset.c, line 554 at r1](https://reviewable.io/reviews/pmem/pmdk/5185#-MW3wwVJ8mNUndfp9vKm:-MX6utVa6kRNFYNNYLmR:b1iyld3) ([raw file](https://github.com/pmem/pmdk/blob/e911f2fb116803d0f9d1cdf2dca3cda166da28aa/src/libpmemset/pmemset.c#L554)):*
<details><summary><i>Previously, kswiecicki (Krzysztof Święcicki) wrote…</i></summary><blockquote>

I omitted resizing/deleting *part_map* at the time and while it didn't leak memory, *ravl* in *pmemset* could contain empty, unmapped part_map records.
*part_map*s should be correctly resized/deleted now.
</blockquote></details>

so it leaked memory, but not in the conventional sense ;)


<!-- Sent from Reviewable.io -->

pbalcer commented:
please rebase, and then we can merge.

kswiecicki commented:
> please rebase, and then we can merge.

Done.

wlemkows reviewed (commented):
> Reviewed 1 of 6 files at r2, 4 of 9 files at r3.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmdk/5185)* status: 7 of 12 files reviewed, 1 unresolved discussion (waiting on @kswiecicki and @pbalcer)



<!-- Sent from Reviewable.io -->

wlemkows reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/pmdk/5185)* status: 7 of 12 files reviewed, 3 unresolved discussions (waiting on @kswiecicki and @pbalcer)

___
*[doc/libpmemset/pmemset_remove_range.3.md, line 40 at r3](https://reviewable.io/reviews/pmem/pmdk/5185#-MYOYJmlEEMKMf2mOXGG:-MYOYJmm3Mcwx2u7C2vw:b5q44wl) ([raw file](https://github.com/pmem/pmdk/blob/904745c34404337bf1667ec8d48b6dd427aa67c5/doc/libpmemset/pmemset_remove_range.3.md#L40)):*
> ```Markdown
> overlapping with virtual memory range described by the *addr* and *len* variables.
> 
> This function will work even if the part mappings contain underlying coalesced parts.
> ```

Ok, but how is it going to work (should be described)
___
*[doc/libpmemset/pmemset_remove_range.3.md, line 58 at r3](https://reviewable.io/reviews/pmem/pmdk/5185#-MYOXu2k6qQqKjYbYpaU:-MYOXu2l0qhHGh4y5WUF:b-blrocb) ([raw file](https://github.com/pmem/pmdk/blob/904745c34404337bf1667ec8d48b6dd427aa67c5/doc/libpmemset/pmemset_remove_range.3.md#L58)):*
> ```Markdown
> # SEE ALSO #
> 
> **pmemset_set_contiguous_part_coalescing(3)**,
> ```

pmem2_map_delete


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[doc/libpmemset/pmemset_remove_range.3.md, line 58 at r3](https://reviewable.io/reviews/pmem/pmdk/5185#-MYOXu2k6qQqKjYbYpaU:-MYOdmkNC6ZFR_MTVazN:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/904745c34404337bf1667ec8d48b6dd427aa67c5/doc/libpmemset/pmemset_remove_range.3.md#L58)):*
<details><summary><i>Previously, wlemkows (Weronika Lewandowska) wrote…</i></summary><blockquote>

pmem2_map_delete
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[doc/libpmemset/pmemset_remove_range.3.md, line 40 at r3](https://reviewable.io/reviews/pmem/pmdk/5185#-MYOYJmlEEMKMf2mOXGG:-MYOdozqBWshIPjqqVn4:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/904745c34404337bf1667ec8d48b6dd427aa67c5/doc/libpmemset/pmemset_remove_range.3.md#L40)):*
<details><summary><i>Previously, wlemkows (Weronika Lewandowska) wrote…</i></summary><blockquote>

Ok, but how is it going to work (should be described)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wlemkows reviewed (commented):
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmdk/5185)* status: 7 of 12 files reviewed, 2 unresolved discussions (waiting on @kswiecicki, @pbalcer, and @wlemkows)



<!-- Sent from Reviewable.io -->

wlemkows reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/pmem/pmdk/5185#-:-MYOpVUt3CUcF_oS9fvf:bnfp4nl)* status: 7 of 12 files reviewed, 2 unresolved discussions (waiting on @kswiecicki, @pbalcer, and @wlemkows)



<!-- Sent from Reviewable.io -->
